src/diagnostic/thread_local.rs

```diff
@@ -57,11 +58,13 @@ impl ThreadLocalDiagnostic {
 }
 
 impl Diagnostic for ThreadLocalDiagnostic {
-    fn visit(&self, visitor: &mut dyn Visitor) -> Result<(), Error> {
+    fn visit<'kvs>(&'kvs self, visitor: &mut dyn Visitor<'kvs>) -> Result<(), Error> {
         CONTEXT.with(|map| {
             let map = map.borrow();
             for (key, value) in map.iter() {
-                visitor.visit(key.into(), value.into())?;
+                let key = key.clone();
+                let value = value.clone();
+                visitor.visit(Key::from(Str::from(key)), Value::from(value))?;
             }
             Ok(())
         })
```

Comment:
Start owned struct from key values.

@KodrAus I wonder if `ValueBag<'a>` can have a `String(String)` variant so that it can be coerced to any lifetime?

Besides, previously the visitor has a method of `fn visit(&mut self, key: Cow<str>, value: Cow<str>) -> Result<(), Error>` so that we don't hit this kind of lifetime hell. Now I'm trying to figure out a way to tell the compiler that the key and value passed here outlive the caller point ..

Reply:
So that we can keep `visitor.visit(key.as_str().into(), value.as_str().into())?;`

Reply:
> I wonder if ValueBag<'a> can have a String(String) variant so that it can be coerced to any lifetime?

Reasonable not to have one since `Clone` over `ValueBag` should be lightweight.

Reply:
Thanks for keeping me looped in @tisonkun. I’m currently on holidays but am keeping the notifications queued so I can look at them properly when I’m back in a week 🙂